internal/pkg/vm/vm.go

```diff
@@ -1,88 +1,60 @@
-func send(ctx context.Context, deps sendDeps, vmCtx *Context) ([][]byte, uint8, error) {
-	if !vmCtx.message.Value.Equal(types.ZeroAttoFIL) {
-		if err := deps.transfer(vmCtx.from, vmCtx.to, vmCtx.message.Value); err != nil {
-			if errors.ShouldRevert(err) {
-				return nil, err.(*errors.RevertError).Code(), err
-			}
-			return nil, 1, err
-		}
-	}
+// GasTracker maintains the state of gas usage throughout the execution of a block and a message
+type GasTracker = gastracker.GasTracker
 
-	if vmCtx.message.Method == types.SendMethodID {
-		// if only tokens are transferred there is no need for a method
-		// this means we can shortcircuit execution
-		return nil, 0, nil
-	}
+// NewGasTracker initializes a new empty gas tracker
+func NewGasTracker() *gastracker.GasTracker {
+	return gastracker.NewGasTracker()
+}
 
-	if vmCtx.message.Method == types.InvalidMethodID {
-		// your test should not be getting here..
-		// Note: this method is not materialized in production but could occur on tests
-		panic("trying to execute fake method on the actual VM, fix test")
-	}
+// NewContextParams is passed to NewVMContext to construct a new context.
+type NewContextParams = vmcontext.NewContextParams
 
-	// TODO: use chain height based protocol version here (#3360)
-	toExecutable, err := vmCtx.actors.GetActorCode(vmCtx.to.Code, 0)
-	if err != nil {
-		return nil, errors.ErrNoActorCode, errors.Errors[errors.ErrNoActorCode]
-	}
+// NewVMContext returns an initialized context.
+func NewVMContext(params NewContextParams) *vmcontext.VMContext {
+	return vmcontext.NewVMContext(params)
+}
 
-	exportedFn, ok := actor.MakeTypedExport(toExecutable, vmCtx.message.Method)
-	if !ok {
-		return nil, 1, errors.Errors[errors.ErrMissingExport]
-	}
+//
+// Free functions
+//
 
-	r, code, err := exportedFn(vmCtx)
-	if r != nil {
-		var rv [][]byte
-		err = encoding.Decode(r, &rv)
-		if err != nil {
-			return nil, 1, errors.NewRevertErrorf("method return doesn't decode as array: %s", err)
-		}
-		return rv, code, err
-	}
-	return nil, code, err
+// Send executes a message pass inside the VM. If error is set it
+// will always satisfy either ShouldRevert() or IsFault().
+func Send(ctx context.Context, vmCtx vmcontext.ExtendedRuntime) ([][]byte, uint8, error) {
+	return vmcontext.Send(ctx, vmCtx)
 }
 
 // Transfer transfers the given value between two actors.
 func Transfer(fromActor, toActor *actor.Actor, value types.AttoFIL) error {
-	if value.IsNegative() {
-		return errors.Errors[errors.ErrCannotTransferNegativeValue]
-	}
-
-	if fromActor.Balance.LessThan(value) {
-		return errors.Errors[errors.ErrInsufficientBalance]
-	}
-
-	fromActor.Balance = fromActor.Balance.Sub(value)
-	toActor.Balance = toActor.Balance.Add(value)
-
-	return nil
+	return vmcontext.Transfer(fromActor, toActor, value)
 }
+
+// FunctionSignature describes the signature of a single function.
+//
+// Dragons: this signature should not be neede on the outside
```

Comment:
I'm not sure I agree with this dragon. It seems reasonable that actors inside the VM should be able to communicate their method signatures to the outside. That will requires something like this type.

Please elaborate or remove.

Reply:
The only things needing to know the signature are whomever is doing encoding the input and decoding the output. This does not necessarily need to exist in data, it could just be code.
The VM runtime doesn't necessarily have to support reflection for someone to be able to call methods on actors and interpret the results correctly. 

We currently send `[]interface{}` down and the `export.go` file is turning it into ABI (this occurs inside the VM). 
The return says `[]byte` which is very unclear on what bytes. We are currently returning ABI bytes which is wrong, there is an issue opened on that already.